sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+                        ResourceId resourceId = metadata.resourceId();
+                        c.output(resourceId);
+                      }
+                    }
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      PCollection<KV<String, byte[]>> files = resourceIds.apply(
+          ParDo.of(
+              new DoFn<ResourceId, KV<String, byte[]>>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  ResourceId resourceId = c.element();
+
+                  try {
+                    ReadableByteChannel channel = FileSystems.open(resourceId);
+                    ByteBuffer byteBuffer = ByteBuffer.allocate(8192);
+                    ByteString byteString = ByteString.EMPTY;
+
+                    while (channel.read(byteBuffer) != -1) {
+                      byteBuffer.flip();
+                      byteString = byteString.concat(ByteString.copyFrom(byteBuffer));
+                      byteBuffer.clear();
+                    }
+
+                    KV<String, byte[]> kv = KV.of(
+                        resourceId.getFilename(),
+                        byteString.toByteArray()
+                    );
+
+                    c.output(kv);
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      return files;
+    }
+  }
+
+  /**
+   * Implements write().
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileBasedSink.convertToFileResourceIfPossible(outputDir));
+    }
+
+    public Write to(ResourceId outputDir) {
+      return toResource(ValueProvider.StaticValueProvider.of(outputDir));
+    }
+
+    public Write toResource(ValueProvider<ResourceId> outputDir) {
+      return toBuilder().setOutputDir(outputDir).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      ResourceId outputDir = getOutputDir().get();
+      if (!outputDir.isDirectory()) {
+        outputDir = outputDir.getCurrentDirectory()
+                             .resolve(outputDir.getFilename(), RESOLVE_DIRECTORY);
+      }
+      final PCollectionView<ResourceId> outputDirView = input.getPipeline()
+          .apply(Create.of(outputDir))
```

Comment:
This is unnecessary, you can access do getOutputDir().get() inside the DoFn rather than passing it as a side input.